[db]  C compilation error

### Describe the bug

C compilation error (from tcc)

### Reproduction Steps

```
module main

import pool
import db.mysql
import db.pg
import time

// ======================================
// 数据库连接配置
// ======================================
pub struct DatabaseConfig {
pub:
	db_type    string // mysql | pgsql | postgres | postgresql
	host       string
	port       u32
	username   string
	password   string
	dbname     string
	ssl_verify bool

	//* pool 配置 */
	max_conns      int           = 100
	min_idle_conns int           = 10
	max_lifetime   time.Duration = 60 * time.minute
	idle_timeout   time.Duration = 30 * time.minute
	get_timeout    time.Duration = 3 * time.second
}

// ======================================
// 统一数据库接口
// ======================================
pub interface DbConnection {
mut:
	query(q string) ![]map[string]string
	execute(q string) !int
	close() !
}

// ======================================
// 数据库连接池 (对业务代码隐藏实现细节)
// ======================================
@[heap]
pub struct DatabasePool {
mut:
	inner &pool.ConnectionPool
}

// ======================================
// PostgreSQL 连接
// ======================================
pub struct PgConnection {
mut:
	conn pg.DB
}

pub fn new_pg_connection(conf DatabaseConfig) !&PgConnection {
	db := pg.connect(pg.Config{
		host:     conf.host
		port:     int(conf.port)
		user:     conf.username
		password: conf.password
		dbname:   conf.dbname
	})!
	return &PgConnection{db}
}

// -------- 核心修复点：Option 解包 ----------
fn get_val(opt ?string) string {
	return opt or { return '' }
}

pub fn (mut c PgConnection) query(q string) ![]map[string]string {
	res := c.conn.exec_result(q)!
	mut rows := []map[string]string{}
	for row in res.rows {
		mut m := map[string]string{}
		for col, idx in res.cols {
			m[col] = get_val(row.vals[idx])
		}
		rows << m
	}
	return rows
}

pub fn (mut c PgConnection) execute(q string) !int {
	res := c.conn.exec(q)!
	return res.len
}

pub fn (mut c PgConnection) close() ! {
	c.conn.close()!
}

// ConnectionPoolable 接口实现
pub fn (mut c PgConnection) validate() !bool {
	c.conn.exec('SELECT 1') or { return error('Postgres validation failed: ${err}') }
	return true
}

pub fn (mut c PgConnection) reset() ! {}

// ======================================
// MySQL 连接
// ======================================
pub struct MysqlConnection {
mut:
	conn mysql.DB
}

pub fn new_mysql_connection(conf DatabaseConfig) !&MysqlConnection {
	db := mysql.connect(mysql.Config{
		host:     conf.host
		port:     conf.port
		username: conf.username
		password: conf.password
		dbname:   conf.dbname
	})!
	return &MysqlConnection{db}
}

pub fn (mut c MysqlConnection) query(q string) ![]map[string]string {
	return c.conn.query(q)!.maps()
}

pub fn (mut c MysqlConnection) execute(q string) !int {
	return c.conn.exec_none(q)
}

pub fn (mut c MysqlConnection) close() ! {
	c.conn.close()!
}

// ConnectionPoolable 接口实现
pub fn (mut c MysqlConnection) validate() !bool {
	c.conn.ping()!
	return true
}

pub fn (mut c MysqlConnection) reset() ! {}

// ======================================
// 连接池管理 (对业务代码透明)
// ======================================
pub fn new_db_pool(conf DatabaseConfig) !&DatabasePool {
	create_conn := fn [conf] () !&pool.ConnectionPoolable {
		match conf.db_type.to_lower() {
			'mysql' {
				return new_mysql_connection(conf)!
			}
			'pgsql', 'postgres', 'postgresql' {
				return new_pg_connection(conf)!
			}
			else {
				return error('Unsupported db type: ${conf.db_type}')
			}
		}
	}

	cp := pool.new_connection_pool(create_conn, pool.ConnectionPoolConfig{
		max_conns:      conf.max_conns
		min_idle_conns: conf.min_idle_conns
		max_lifetime:   conf.max_lifetime
		idle_timeout:   conf.idle_timeout
		get_timeout:    conf.get_timeout
	})!

	return &DatabasePool{cp}
}

// 业务友好的接口
pub fn (mut p DatabasePool) acquire() !DbConnection {
	mut conn := p.inner.get()!
	return conn as DbConnection
}

pub fn (mut p DatabasePool) release(conn DbConnection) ! {
	p.inner.put(conn as &pool.ConnectionPoolable)!
}

pub fn (mut p DatabasePool) close() {
	p.inner.close()
}

// ======================================
// 业务代码使用示例 (简洁明了) (完全不需要关心具体数据库类型)
// ======================================
fn main() {
	// 配置 (可以来自配置文件)
	conf := DatabaseConfig{
		db_type:  'mysql' // 只需改这里就可以切换数据库
		host:     '127.0.0.1'
		port:     3306
		username: 'root'
		password: 'mysql_123456'
		dbname:   'vcore'
	}

	// 创建连接池 (初始化代码)
	mut db_pool := new_db_pool(conf) or { panic(err) }
	defer {
		db_pool.close()
	}

	// 获取连接
	mut conn := db_pool.acquire() or { panic(err) }
	defer {
		db_pool.release(conn) or { println('Warning: failed to return connection') }
	}

	// 执行查询 - 业务代码只关心业务逻辑
	users := conn.query('SELECT * FROM sys_user') or { panic(err) }
	println('User: ${users}')

	// 执行更新
	// affected := conn.execute('UPDATE users SET last_login = NOW() WHERE id = 1') or { panic(err) }
	// println('Updated ${affected} rows')

	// 事务示例 (伪代码)
	// conn.execute('BEGIN')!
	// // ... 业务操作
	// conn.execute('COMMIT')!

	println('All operations completed successfully!')
}

```
v  run   .


### Expected Behavior

v panic

### Current Behavior

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/dbp.01KAFZQGJH7XJC8C1TZ2Y9NYSG.tmp.c:15885: warning: assignment discards qualifiers from pointer target type
cc: /tmp/v_1000/dbp.01KAFZQGJH7XJC8C1TZ2Y9NYSG.tmp.c:16649: error: cannot convert 'struct pool__ConnectionPoolable' to 'struct pool__ConnectionPoolable *'
...
cc: /tmp/v_1000/dbp.01KAFZQGJH7XJC8C1TZ2Y9NYSG.tmp.c:16649: error: cannot convert 'struct pool__ConnectionPoolable' to 'struct pool__ConnectionPoolable *'
(note: the original output was 5 lines long; it was truncated to its first 2 lines + the last line)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.12 2a12104

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.12 c560e7290059aaff6d43005c5bc5d0ed9dab4d9f.2a12104
|:-------------------|:-------------------
|OS                  |linux, Deepin 25
|Processor           |6 cpus, 64bit, little endian, Intel(R) Core(TM) i5-8500B CPU @ 3.00GHz
|Memory              |0.31GB/7.6GB
|                    |
|V executable        |/home/Jengro/.vmr/versions/v_versions/v_latest/v
|V last modified time|2025-11-20 01:04:11
|                    |
|V home dir          |OK, value: /home/Jengro/.vmr/versions/v_versions/v_latest
|VMODULES            |OK, value: /home/Jengro/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/Jengro/Documents/Dev/v-product/vprod-workspase/backend/adapter/dbpool
|                    |
|Git version         |git version 2.50.1
|V git status        |weekly.2025.45-90-g2a121047
|.git/config present |true
|                    |
|cc version          |cc (Deepin 12.3.0-17deepin15) 12.3.0
|gcc version         |gcc (Deepin 12.3.0-17deepin15) 12.3.0
|clang version       |Deepin clang version 17.0.6 (5deepin6)
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (Debian GLIBC 2.38-6deepin17) 2.38
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -12627,12 +12627,19 @@ fn (mut g Gen) as_cast(node ast.AsCast) {
 			g.as_cast_type_names[idx] = variant_sym.name
 		}
 	} else if expr_type_sym.kind == .interface && sym.kind == .interface {
+		is_ptr_target := node.typ.is_ptr()
+		if is_ptr_target {
+			g.write('HEAP(${sym.cname}, ')
+		}
 		g.write('I_${expr_type_sym.cname}_as_I_${sym.cname}(')
 		if node.expr_type.is_ptr() {
 			g.write('*')
 		}
 		g.expr(node.expr)
 		g.write(')')
+		if is_ptr_target {
+			g.write(')')
+		}
 
 		mut info := expr_type_sym.info as ast.Interface
 		right_info := sym.info as ast.Interface

--- a/vlib/v/tests/interfaces/interface_to_ptr_interface_as_cast_test.v
+++ b/vlib/v/tests/interfaces/interface_to_ptr_interface_as_cast_test.v
@@ -1,0 +1,46 @@
+// regression test for https://github.com/vlang/v/issues/25785
+// `iface as &OtherIface` must produce a valid pointer to the converted interface.
+
+pub interface Poolable {
+mut:
+	validate() !bool
+	close() !
+	reset() !
+}
+
+pub interface DbConn {
+mut:
+	query(q string) ![]string
+	close() !
+}
+
+pub struct PgConn {
+pub mut:
+	tag string
+}
+
+pub fn (mut c PgConn) query(q string) ![]string {
+	return [c.tag, q]
+}
+
+pub fn (mut c PgConn) close() ! {}
+
+pub fn (mut c PgConn) validate() !bool {
+	return true
+}
+
+pub fn (mut c PgConn) reset() ! {}
+
+fn take_ptr(p &Poolable) string {
+	mut conn := unsafe { p }
+	return if conn.validate() or { false } { 'ok' } else { 'bad' }
+}
+
+fn iface_to_ptr_iface(c DbConn) string {
+	return take_ptr(c as &Poolable)
+}
+
+fn test_iface_as_ptr_iface() {
+	c := DbConn(PgConn{ tag: 'pg' })
+	assert iface_to_ptr_iface(c) == 'ok'
+}

